cupy/core/core.pyx

```diff
@@ -1834,6 +1840,28 @@ cdef _take_kernel_0axis = ElementwiseKernel(
     'cupy_take_0axis')
 
 
+cdef _boolean_array_indexing_nth = ElementwiseKernel(
+    'T a, bool boolean_array, S nth',
+    'raw T out',
+    'if (boolean_array) out[nth] = a',
+    'cupy_boolean_array_indexing_nth')
+
+
+cpdef ndarray _boolean_array_indexing(ndarray a, ndarray boolean_array):
+    cdef int i, n_true
+    nth = ndarray(boolean_array.shape, dtype=int)
+    n_true = 0
+    boolean_array = boolean_array.flatten()
+    for i in range(boolean_array.size):
```

Comment:
This implementation loops over the whole array, which takes O(array_size). I think [`scan`](https://github.com/pfnet/chainer/blob/master/cupy/core/core.pyx#L2563) is useful implement this feature. So could you try to improve the implementation?